src/transform/src/attribute/cardinality.rs

```diff
@@ -0,0 +1,646 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Definition and helper structs for the [`Cardinality`] attribute.
+
+use std::collections::{BTreeMap, BTreeSet};
+
+use mz_expr::{
+    BinaryFunc, Id, JoinImplementation, MirRelationExpr, MirScalarExpr, TableFunc, UnaryFunc,
+    VariadicFunc,
+};
+use mz_ore::cast::CastLossy;
+use mz_repr::GlobalId;
+
+use mz_repr::explain::ExprHumanizer;
+use ordered_float::OrderedFloat;
+
+use crate::attribute::subtree_size::SubtreeSize;
+use crate::attribute::unique_keys::UniqueKeys;
+use crate::attribute::{Attribute, DerivedAttributes, Env, RequiredAttributes};
+use crate::symbolic::SymbolicExpression;
+
+use super::Arity;
+
+/// Compute the estimated cardinality of each subtree of a [MirRelationExpr] from the bottom up.
+#[allow(missing_debug_implementations)]
+pub struct Cardinality {
+    /// Environment of computed values for this attribute
+    env: Env<Self>,
+    /// A vector of results for all nodes in the visited tree in
+    /// post-visit order
+    pub results: Vec<SymExp>,
+    /// A factorizer for generating appropriating scaling factors
+    pub factorize: Box<dyn Factorizer + Send + Sync>,
+}
+
+impl Default for Cardinality {
+    fn default() -> Self {
+        Cardinality {
+            env: Env::default(),
+            results: Vec::new(),
+            factorize: Box::new(WorstCaseFactorizer {
+                cardinalities: BTreeMap::new(),
+            }),
+        }
+    }
+}
+
+/// The variables used in symbolic expressions representing cardinality
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum FactorizerVariable {
+    /// The total cardinality of a given global id
+    Id(GlobalId),
+    /// The inverse of the number of distinct keys in the index held in the given column, i.e., 1/|# of distinct keys|
```

Comment:
Oh, this might be what we need to implement that formula from the paper. Do we actually have this info from somewhere, or is this just a placeholder for now?

Reply:
It's a placeholder---storage/persist says getting this info will be a little tricky.